feat: spec decode with draft models

tomasruizt opened the pull request:
## Purpose
Enabling draft models for speculative decoding (SD).
E.g. `Qwen3-1.7B` as draft model and `Qwen3-32B` as target model.
This type of SD requires no special trained heads (like EAGLE, or Medusa).

Example usage:
```shell
vllm serve \
    --model=Qwen/Qwen3-4B \
    --speculative-config '{"model": "Qwen/Qwen3-0.6B", "method": "draft_model", "num_speculative_tokens": 3, "max-model-len": 2000, "disable_padded_drafter_batch": true}' \
    --max-model-len 2000
```
Get a generation:

```shell
curl -X POST http://localhost:8000/v1/completions \
    -H "Content-Type: application/json" \
    -d '{"prompt": "Capital of France", "max_tokens": 16}'
```
## Status
* The acceptance rates in greedy decoding are great for Qwen3 models (see corresponding section).
* Using SD with Qwen3 has higher throughput (TPOT) than not using SD.

## Acceptance Length
As suggested by @ekagra-ranjan, I benchmarked acceptance length (AL) with the command below:
```shell
VLLM_USE_V1=1 python3 examples/offline_inference/spec_decode.py \
    --model-dir Qwen/Qwen3-32B \
    --draft-model Qwen/Qwen3-1.7B \
    --method draft_model \
    --num_spec_tokens 3 \
    --dataset-name hf \
    --dataset-path philschmid/mt-bench \
    --num_prompts 100 \
    --temp 1.0 \
    --gpu-memory-utilization 0.9
```

The AL values within the Qwen3 family seem good, both with temperatures of 0.0 (greedy) and 1.0.
As a sanity check, I benchmarked LLama-3.2-1B as both target and draft, which had almost perfect AL (3.97/4), suggesting its working as intended.
I have not run the default model `meta-llama/Llama-3.1-8B-Instruct`, because I didn't find a good draft model for it, but feel free to suggest one and I can run the benchmarks.


Temperature t=0:
| Target | Draft | K | Temperature | AL |
|---|---|---|---|---|
| Qwen3-32B | Qwen3-1.7B | 3 | 0.0 | 2.79 |
| Qwen3-32B | Qwen3-1.7B | 4 | 0.0 | 3.12 |
| Llama-3.2-1B | Llama-3.2-1B | 3 | 0.0 | 3.97 |

Temperature t=1.0:
| Target | Draft | K | Temperature | AL |
|---|---|---|---|---|
| Qwen3-32B | Qwen3-1.7B | 3 | 1.0 | 2.61 |
| Qwen3-32B | Qwen3-1.7B | 4 | 1.0 | 2.85 |
| Llama-3.2-1B | Llama-3.2-1B | 3 | 1.0 | 2.82 |

Using t=1.0, the AL metric degrades. However, spec-decode with probabilities is not yet implemented, needed for lossless rejection sampling. This is being addressed atm: https://github.com/vllm-project/vllm/pull/20459. After that PR is merged, the AL for non-greedy spec-decode should improve.

All scripts and logs used for the benchmarks can be found in [this Google Drive](https://drive.google.com/drive/folders/1mFgeIVUI16VMlXJe8jm2TImY0lYKSQLC?usp=drive_link).

## Online Throughput Metrics

I measured online throughput metrics using the commands below. Hardware was an RTX PRO 6000 96GB. After making sure the draft model also uses CUDA graph, SD has higher throughput than not using SD. See tables below.

```shell
VLLM_USE_V1=1 vllm serve Qwen/Qwen3-32B \
  --max-model-len 20000 \
  --disable-uvicorn-access-log

# or 

VLLM_USE_V1=1 vllm serve Qwen/Qwen3-32B \
  --speculative_config '{"method": "draft_model", "model": "Qwen/Qwen3-1.7B", "num_speculative_tokens": 3, "max_model_len": 20000, "disable_padded_drafter_batch": true}' \
  --max-model-len 20000 \
  --disable-uvicorn-access-log

nohup vllm bench serve \
  --model Qwen/Qwen3-32B \
  --dataset-name hf \
  --dataset-path philschmid/mt-bench \
  --num-prompts 80 \
  --max-concurrency (100|1) \
  --temperature 0.0 \
  --top-p 1.0 2>&1 > results/qwen3-32b-t0-run1.out &
```

* The tables show shorter runtime and higher throughput for SD (both in batch size 1 and 100). 
* Using SD the TPOT is 50% shorter (better) in batch size 1, and 26% to 33% shorter in batch size 100. The reason is the higher throughput of the draft model.
* Using SD the TTFT and ITL are higher (worse), because tokens are produced in batches by the spec-decoding. Nevertheless, total runtimes are shorter overall when using SD.

The metrics (lower is better) are:
* TTFT: Time-to-first-token
* TPOT: Time-per-output-token
* ITL: Inter-token-latency

<details>
<summary>Batch Size = 1</summary>
For Temperature = 0.0:

| Target | Draft | Runtime |  TTFT | TPOT | ITL |
|---|---|---|---|---|---|
| Qwen3-32B | - | 943s | 69.14ms | 46.06ms | 45.88ms |
| Qwen3-32B | Qwen3-1.7B | 466s | 76.77ms | 22.62ms | 62.44ms |

Using SD runtimes and TPOT are shorter by ~50%.
</details>

<details>
<summary>Batch Size = 100</summary>
For Temperature = 0.0:

| Target | Draft | Runtime |  TTFT | TPOT | ITL |
|---|---|---|---|---|---|
| Qwen3-32B | - | 16.88s | 262.59ms | 65.09ms | 64.84ms |
| Qwen3-32B | Qwen3-1.7B | 13.04s | 284.45ms | 43.70ms | 121.48ms |

For Temperature = 1.0:

| Target | Draft | Runtime |  TTFT | TPOT | ITL |
|---|---|---|---|---|---|
| Qwen3-32B | - | 16.83s | 230.04ms | 64.95ms | 64.70ms |
| Qwen3-32B | Qwen3-1.7B | 14.84s | 272.51ms | 48.00ms | 122.27ms |
</details>

This scenario with batch size 100 is a more realistic inference case.
Using SD runtimes and TPOT are shorter.
### Profiling

This section was removed, since using CUDA graphs on the draft model significantly improved its speed.

<details>
<summary>Profiling script</summary>
I used the command below to profile the generation process and identify that the draft model was running too slow before.

```shell
export VLLM_USE_V1=1
export VLLM_TORCH_PROFILER_DIR=./profiles/
export CUDA_LAUNCH_BLOCKING=1

vllm bench throughput \
    --model=Qwen/Qwen3-32B \
    --speculative-config '{"model": "Qwen/Qwen3-1.7B", "method": "draft_model", "num_speculative_tokens": 3, "max_model_len": 2048}' \
    --dataset-name=hf \
    --dataset-path=likaixin/InstructCoder \
    --max-num-seqs=100 \
    --num-prompts=10 \
    --input-len=1000 \
    --output-len=10 \
    --max-model-len=2048 \
    --gpu-memory-utilization=0.95 \
    --profile
```
Note: The command uses the `--profile` flag, which I introduce in this PR: https://github.com/vllm-project/vllm/pull/24575
</details>

## Test Plan
The added unit test check the correctness metrics. To run it:
```shell
cd tests/v1/e2e/
pytest test_spec_decode.py -k test_draft_model_correctness
```

<details>

<summary>EAGLE testing</summary>

I tested that the EAGLE implementation stays unaffected the command below

```shell
VLLM_USE_V1=1 python3 examples/offline_inference/spec_decode.py \
    --model-dir meta-llama/Llama-3.1-8B-Instruct \
    --eagle-dir yuhuili/EAGLE3-LLaMA3.1-Instruct-8B \
    --method eagle3 \
    --num_spec_tokens 7 \
    --dataset-name hf \
    --dataset-path philschmid/mt-bench \
    --num_prompts 80 \
    --temp 0.0 \
    --gpu-memory-utilization 0.9
```
The results are in line with previous measurements like https://github.com/vllm-project/vllm/pull/17504#issuecomment-2845707371

```shell
total_num_output_tokens: 16990
num_drafts: 4816
num_draft_tokens: 33712
num_accepted_tokens: 12208
mean acceptance length: 3.53
--------------------------------------------------
acceptance at token 0: 0.74
acceptance at token 1: 0.54
acceptance at token 2: 0.41
acceptance at token 3: 0.31
acceptance at token 4: 0.23
acceptance at token 5: 0.17
acceptance at token 6: 0.13
```

</details>


## Follow-up Optimizations
* Include the tokens in `next_token_ids` together with `target_token_ids` in the first forward pass of the draft model. This reduces the number of forward passes needed in each drafting phase by one, speeding up drafting.

# Qwen3 Metrics 
Note: The code for the visualizations below can be found in this gist: https://gist.github.com/tomasruizt/81aa1dee4fed88bfad2fe746c3f66739

I compare Qwen3-32B against Qwen3-32B with Qwen3-1.7B as drafter model, on a single H100 GPU (TP=1).
The benchmarks show (left to right)
* An increase in Total Token Throughput with 1 to 100 concurrent requests (so larger batch sizes).
* A decrease Benchmark Duration
* Increases in Total Token Throughput of almost 2x

<img width="31%" alt="image" src="https://github.com/user-attachments/assets/6316dc9e-488d-4170-877b-4207baee0db5" />

<img width="31%" alt="image" src="https://github.com/user-attachments/assets/b6a93069-f573-4764-9f96-084e84881a18" />

<img width="31%" alt="image" src="https://github.com/user-attachments/assets/572b97a8-d2b0-463c-9583-c4a555595c6d" />

Broken down we find:
* A steep improvement (reduction) of Time Per Output Token (TPOT)
* Higher Inter-Token-Latency (ITL), because tokens are generated and verified in batches
* Equivalent Time-To-First-Token (TTFT) except in high batch sizes.

<img width="93%" alt="image" src="https://github.com/user-attachments/assets/28a60a9c-90ad-4e29-855b-71b317546070" />

# Llama3 Multi-GPU Metrics
I benchmarked `meta-llama/Meta-Llama-3-70B`, with `meta-llama/Llama-3.2-1B` as a draft model on 4 x H100 GPUs (TP=4). The goal was to measure the effect of tensor parallelism on throughput, since the small draft model is also running on the same level of tensor parallelism as the target model.  I found an acceleration in this setup only up to bsz=32, and a slowdown for bsz=64. This setup shows that using TP > 1 degrades the acceleration of draft_model. Presumably, because of the inter-GPU communication overhead for the draft model. In a follow-up PR this should be optimized.

[serving-script.sh](https://github.com/user-attachments/files/23006446/serving-script.sh)
[benchmark-script.sh](https://github.com/user-attachments/files/23006447/benchmark-script.sh)

<img width="31%" alt="image" src="https://github.com/user-attachments/assets/f125d957-fe17-4995-aba6-b8a3990ce4c6" />
<img width="31%" alt="image" src="https://github.com/user-attachments/assets/6488414f-a119-4e2b-90a1-0dbb5ad661fa" />
<img width="31%" alt="image" src="https://github.com/user-attachments/assets/4f2c4dcd-4751-478e-9b3e-bfcce21aa6f4" />

The TPOT metric improves in general, showing similar results to Qwen3.

<img width="93%" alt="image" src="https://github.com/user-attachments/assets/19789f11-4f73-41ba-a33b-c0ecf6196c99" />

# EAGLE3 Metrics (Reference)
Below are benchmark values using method="eagle3", with target model `meta-llama/Llama-3.3-70B-Instruct` and draft model `yuhuili/EAGLE3-LLaMA3.3-Instruct-70B` on 4 x H100 GPUs (TP=4). Eagle achieves over 2x acceleration in batch_size=1. In contrast to draft_model, the eagle drafter is running with `tensor_parallelism = 1`.

[serve-script.sh](https://github.com/user-attachments/files/23017732/serve-script.sh)
[bench-script.sh](https://github.com/user-attachments/files/23017733/bench-script.sh)

<img width="31%" alt="image" src="https://github.com/user-attachments/assets/eabe9097-511e-42ab-9cfd-d14755fba842" />
<img width="31%" alt="image" src="https://github.com/user-attachments/assets/b4d996e6-b193-463f-a9c0-eaf95b2321d0" />
<img width="31%" alt="image" src="https://github.com/user-attachments/assets/e4aec257-5d6f-4abb-91a8-29b7e7e0484d" />

<img width="93%" alt="image" src="https://github.com/user-attachments/assets/9f48750f-3753-4de3-95f5-c384f05708dc" />

---

> [!NOTE]
> Introduces draft-model-based speculative decoding and wires it across configs, runtime, and tests.
> 
> - Adds `DraftModelProposer` with Triton `merge_toks_kernel`, `extend_all_queries_by_1`, and padded-batch path; integrates with `gpu_model_runner`, scheduler lookahead, and K/V cache binding (interleaving target/draft layers)
> - New config capabilities/validation: `SpeculativeConfig.uses_draft_model()`, vocab-size check, explicit `draft_tensor_parallel_size` (rejects `tensor_parallel_size`), per-draft `ParallelConfig` creation, `disable_logprobs`, and `num_lookahead_slots`; `ParallelConfig.replace()` and `VllmConfig.replace()` helpers
> - Model loading updated to accept `prefix` and separate draft `VllmConfig`; enforce TP constraints and isolate quantization between target/draft
> - Metrics helpers to compute acceptance rate/length; ready checker logs when endpoint isn’t ready
> - Example script adds `--method draft_model`, `--draft-model`, GPU mem utilization, and max seqs; passes new speculative fields
> - Extensive tests: correctness/quantization/TP for draft-model SD, engine-args validation, KV cache binding with draft layers, and kernel unit tests
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d9641ad89c71712b35ab9fdc05777ead16818b56. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Enables speculative decoding using a separate draft model and integrates it across configuration, runtime, and tests.
> 
> - New `DraftModelProposer` with Triton `merge_toks_kernel` and query-extension path; integrates with `gpu_model_runner`, scheduler lookahead, padded-batch path, and interleaved K/V cache binding
> - Config updates: `SpeculativeConfig.uses_draft_model()`, vocab-size validation, explicit `draft_tensor_parallel_size` (rejects `tensor_parallel_size`), draft `ParallelConfig` creation, and TP constraints; add `ParallelConfig.replace()` and `VllmConfig.replace()`
> - Model loading supports draft-specific `VllmConfig` and `prefix` to namespace modules; isolates draft quantization/TP from target
> - Attention utils: `extend_all_queries_by_1`, `CommonAttentionMetadata` helpers; scheduler now sets lookahead for draft-model SD
> - Metrics helpers to compute acceptance rate/length; ready-checker logs when endpoint isn’t ready
> - Example `spec_decode.py` adds `--method draft_model`, `--draft-model`, GPU mem/max seqs flags; forwards new speculative fields
> - Tests: correctness/quantization/TP for draft-model SD, engine-args validation, KV cache binding for draft layers, and kernel unit tests
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d708778513f8b3cceb3f93ce6e08c77661c14bce. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces draft-model-based speculative decoding and wires it across config, runtime, and tests.
> 
> - New `DraftModelProposer` with Triton `merge_toks_kernel` and query-extension (`extend_all_queries_by_1`); integrates with scheduler lookahead and `gpu_model_runner` (padded-batch path, GPU tokens) and validates KV cache grouping
> - Config: add `uses_draft_model()`, vocab-size check, explicit `draft_tensor_parallel_size` (rejects `tensor_parallel_size`), per-draft `ParallelConfig`/`VllmConfig` via `create_vllm_config_for_draft_model`, and `replace()` helpers for `ParallelConfig`/`VllmConfig`
> - Model loading supports draft-specific configs and `prefix` to namespace modules across loaders (base, GGUF, tensorizer)
> - Worker/utils: bind K/V caches interleaving target and draft layers; scheduler sets `num_lookahead_tokens` for draft-model SD
> - Bench/UX: ready-checker warns when endpoint not ready; example script adds `--method draft_model`, `--draft-model`, memory/max-seqs flags
> - Metrics: helpers to compute acceptance rate/length
> - Tests: correctness/quantization/TP for draft-model SD, engine-arg validation, KV-cache binding, and kernel unit tests
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 5ee3a595a1dade9e062f90b55e14010983553aeb. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

ekagra-ranjan commented:
@tomasruizt - Thank you for the PR! 
1. Can you also report Acceptance Length (AL) and K used? It is more informative than AR. For e.g., with Eagle with K=3 we get AL ~2.29 on MTBench so we can expect 2.29x speedup assuming 0 draft overhead. AL gives a more holistic overview of the speedup for a given K than AR. 
2. Can you run the metrics on MTBench instead of one off sample for less noisy metric? This would also allow us to compare the AL and TPOT improvement from other SD methods: https://github.com/vllm-project/vllm/issues/17812

tomasruizt commented on vllm/engine/arg_utils.py:
```diff
@@ -1486,10 +1486,7 @@ def _is_v1_supported_oracle(self, model_config: ModelConfig) -> bool:
         if (self.speculative_config is not None
                 and self.speculative_config.get("method") == "draft_model"):
-            raise NotImplementedError(
-                "Speculative decoding with draft model is not supported yet. "
-                "Please consider using other speculative decoding methods "
-                "such as ngram, medusa, eagle, or deepseek_mtp.")
+            return True
```
Makes sense!

tomasruizt commented on vllm/v1/worker/utils.py:
```diff
@@ -255,25 +254,11 @@ def bind_kv_cache(
         # NOTE: Use list because of v0 PP virtual engine.
-        forward_context[layer_name].kv_cache = [kv_cache]
+        forward_context[layer].kv_cache = [kv_caches[layer]]
+        runner_kv_caches.append(kv_caches[layer])
```
Can you reevaluate your review?

tomasruizt commented on vllm/v1/spec_decode/eagle.py:
```diff
@@ -318,22 +350,44 @@ def propose(
+            batch_descriptor = BatchDescriptor(num_tokens=input_batch_size,
+                                               uniform_decode=True)
+            cudagraph_runtime_mode, batch_descriptor = \
+                self.runner.cudagraph_dispatcher.dispatch(batch_descriptor)
```
Yes, `batch_descriptor` and `cudagraph_runtime_mode` need to be passed to s`et_forward_context()` to replay the previously captured CUDA graph.

ekagra-ranjan commented:
@tomasruizt - 
1. The online inference has this flag `--max-concurrency 100`. This means that you are testing on Batch size of 100. Please use 1 instead of 100 to iterate as of now.  When things are finalized then you can provide numbers for BS=1/4/16. 
2. Can you also provide the cmd you used to profile the model for reproducibility?
3. Curious what change you did to make the draft model use cugraph when it was not using earlier?

tomasruizt commented:
@ekagra-ranjan
* I've added throughput metrics with batch size = 1. 
  * With batch_size=1, using SD reduces runtime and TPOT by 50%. I.e. inference is 2x as fast 🚀 
  * With batch_size=100, using SD also reduces runtime and TPOT, by 26% to 33%. This batch size gives a better impression of SD performance under a more realistic and challenging workload.
* I've added the profiling command in the **Profiling** section

LiuXiaoxuanPKU commented:
Hi @tomasruizt , I discussed with @WoosukKwon last week about this PR, the major concern is code complexity.
Questions before we move on:
1. Just want to double check, if the draft model has different `num_head` and `head_dim`, then we cannot support it. --> This is fine, we just need to document it somewhere.
2. How do you handle the KV cache for the draft model, say we propose 5 tokens, only 3 of them are accepted. 

Code complexity related changes:
1. Could you separate this PRs into at least two? One is the draft model based spec decode functionality, the other is cuda graph support for the draft model.
2. Could you separate draft model proposer logic and eagle proposer logic? If they share some common logic, you can create some abstraction, but it's also ok if we don't do it for now.

benchislett commented on vllm/v1/spec_decode/eagle.py:
```diff
@@ -742,3 +669,97 @@ def compute_probs_and_sample_next_token(
     return next_token_ids, probs
+
+
+def eagle_prepare_inputs(
```
this should probably be `drafter_prepare_inputs` if it's not specific to EAGLE

benchislett commented on vllm/v1/spec_decode/draft_model.py:
```diff
@@ -0,0 +1,256 @@
+        ):
+            self.model(**model_kwargs)
+
+    # Copied and adapted from eagle.py
```
I think the 'proper' way to handle this is to factor out as much of the common components as possible, so that what remains is unique to the backend. In practice though this might not be so easy.

benchislett replied:
Could you explain more about this `n_forward_passes` concept? Why do you need an additional forward pass?

benchislett replied:
I do not currently see the issue in doing the prefill "properly". I understand that the EAGLE speculator does it differently, but it is not clear to me what the obstacle is. Is there a problem with KV allocation being insufficient for a draft-model? Or is it just a matter of rewriting `prepare_inputs` for the drafter to handle the difference in prompt token size?

tomasruizt replied:
In this PR I introduce the tests `test_draft_model_correctnes()`, which don't exist on main.
Assuming I submit the refactoring PR first, which is a best practice, I would not have the new tests to make sure the base class I extract also supports the draft-model implementation. That is what I meant.

tomasruizt commented:
I had made a mistake while merging main to this branch, and the bot added a bunch of extra labels. But the commits look Ok again.

tomasruizt commented:
@LiuXiaoxuanPKU I separated the proposing logic of EAGLE and Draft Models and they share a superclass for common logic. The CUDA graph code is still in this PR because the code changes are small and it speeds up the draft model. You suggested splitting the CUDA graph code into a separate PR. Was this to reduce code complexity, or rather to establish before & after performance? I can still break out the CUDA graph code into a separate PR, just let me know again.

tomasruizt commented:
After splitting the `DraftModelProposer` from the `EagleProposer` I reran the online throughput metrics for batch size 1, as suggested by @ekagra-ranjan, this time on an H100. Spec decoding still improves throughput by 2x for Qwen3-32B (batch_size=1). The command is the same as in the PR description. Results are [here](https://drive.google.com/drive/folders/1lei5BK63WwPuSY-LLR6gpATjuyQRpTVx?usp=drive_link), including TTFT, TPOT, and ITL.

benchislett reviewed (changes requested):
I am not comfortable with the current state of this PR. There are a few main issues I see:
- Draft model running an additional decode instead of running a differently-sized prefill. This seems like a solvable problem and an obstacle to both performance as well as code reusability between the EAGLE and draft-model pathways.
- There is still a lot of duplicate code in this PR, especially in the propose method. This will be difficult to maintain in its current state. I think that if we are going to merge support for draft-model spec decode then code quality should be a baseline that we can build up from instead of an objective for follow-up work.

tomasruizt commented:
@benchislett 

As we discussed in our call, I moved the `propose()` method into the superclass so it could be reused among classes. I also introduced flags to the constructor to toggle the behavior between EAGLE and DraftModel:
1. `pass_hidden_states_to_model: bool` True for EAGLE
2. `pass_cudagraph_args_to_forward_ctx: bool` True for DraftModel
3. `drop_first_drafted_tokens: bool` True for DraftModel
4. A variable `self.num_forward_passes` which is 1 larger for the DraftModel

Number 3 and 4 will become unnecessary if we implement the optimal prefill we discussed, which would reduce the forward passes by one, and improve runtimes. Nevertheless, it's useful to remember the major factor affecting DraftModel speed is the CUDA graph usage, which is now conveniently a single flag.

At the moment the EAGLE file diff looks horrible, I guess because of the combination of extracting a superclass plus introducing flags to the constructor. Let me know if you would prefer to review the EAGLE refactor as a separate PR to main (or as multiple small PRs) 👍

tomasruizt commented:
The EAGLE code is frequently changed on main, so it is difficult to move EAGLE code around without painful merge conflicts.
To get nice git diffs on the EAGLE file I put the body of EAGLE into the superclass.

tomasruizt commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -281,6 +282,12 @@ def __init__(
         if self.speculative_config and get_pp_group().is_last_rank:
             if self.speculative_config.method == "ngram":
                 self.drafter = NgramProposer(self.vllm_config)
+            elif self.speculative_config.uses_draft_model():
+                self.drafter = DraftModelProposer(
+                    vllm_config=self.vllm_config,
+                    device=self.device,
+                    runner=self,
+                )  # type: ignore
```
I haven't addressed this use case yet. Would you mind sharing the vllm command you are using? @smsegal

tomasruizt replied:
@smsegal I tested putting the draft model into a single rank, but it was complicated. This would be a very valuable follow-up optimization

tomasruizt commented:
I refactored to remove the extra forward pass. However, the new acceptance length (AL) I'm measuring atm  is lower than before the refactor. This could be related to this recent bug in `main` with AL: https://github.com/vllm-project/vllm/issues/26191
I want to solve that first before making statements about AL with/without the extra forward pass.

tomasruizt commented:
After refactoring I've benchmarked acceptance length (AL) and throughput metrics. The AL remains at the same value as before the refactor (2.79). This requires `disable_padded_drafter_batch = True` to avoid the AL regression documented here: https://github.com/vllm-project/vllm/issues/26191.

<details>
<summary>AL command</summary>
#!/bin/bash
#SBATCH --partition lrz-hgx-h100-94x4
#SBATCH --gres=gpu:1
#SBATCH --array=0
#SBATCH -o /dss/dssfs02/lwp-dss-0001/pn76je/pn76je-dss-0000/vllm/logs/%x/%a_out.logs
#SBATCH -e /dss/dssfs02/lwp-dss-0001/pn76je/pn76je-dss-0000/vllm/logs/%x/%a_err.logs
#SBATCH --job-name vllm-acceptance-len-t0.0
#SBATCH --time 00:30:00

# Set container name based on job name and array task ID to avoid conflicts
CONTAINER_NAME=vllm-acceptance-len-t0.0

# Delete any existing container with this name
echo "Removing any existing container named $CONTAINER_NAME..."
enroot remove --force $CONTAINER_NAME || true

# Create new container
echo "Creating new container $CONTAINER_NAME..."
enroot create --name $CONTAINER_NAME /dss/dssfs02/lwp-dss-0001/pn76je/pn76je-dss-0000/tomasruiz/my_custom_pt.sqsh

# Start the container with necessary mounts and environment variables
enroot start --rw --root \
    --mount /dss/dsshome1/0D/di38bec/code:/workspace/code \
    --mount /dss/dsshome1/0D/di38bec/datasets:/workspace/datasets \
    --mount "$DSS_HOME:$DSS_HOME" \
    --env DSS_HOME=$DSS_HOME \
    --env HF_TOKEN=$HF_TOKEN \
    --env HF_HOME=$HF_HOME \
    --env VLLM_USE_V1=1 \
    $CONTAINER_NAME bash -c "source /root/ptvenv/bin/activate && python3 /workspace/code/vllm/examples/offline_inference/spec_decode.py \
        --model-dir Qwen/Qwen3-32B \
        --draft-model Qwen/Qwen3-1.7B \
        --method draft_model \
        --num_spec_tokens 3 \
        --dataset-name hf \
        --dataset-path philschmid/mt-bench \
        --num_prompts 100 \
        --temp 0.0 \
        --gpu-memory-utilization 0.9"

# Clean up: remove the container after job completion
echo "Cleaning up: removing container $CONTAINER_NAME..."
enroot remove --force $CONTAINER_NAME || true
</details>

In terms of throughput metrics, I'm using the `mtbench` dataset with several levels of concurrency, which shows the performance of spec decode at different batch sizes. The performance metrics are good, with total token throughput of Qwen3-32B increasing between 40% and 90%.

<details>
<summary>Throughput Benchmark Command</summary>
#!/bin/bash
#SBATCH --partition lrz-hgx-h100-94x4
#SBATCH --gres=gpu:1
#SBATCH --array=0
#SBATCH -o /dss/dssfs02/lwp-dss-0001/pn76je/pn76je-dss-0000/vllm/logs/%x/%a_out.logs
#SBATCH -e /dss/dssfs02/lwp-dss-0001/pn76je/pn76je-dss-0000/vllm/logs/%x/%a_err.logs
#SBATCH --job-name vllm-throughput-sd-c1-t0.0
#SBATCH --time 00:30:00

# Set container name based on job name and array task ID to avoid conflicts
CONTAINER_NAME=vllm-throughput-sd-c1-t0.0

# Calculate port based on base port and array task ID to avoid conflicts
VLLM_PORT=15000

# Delete any existing container with this name
echo "Removing any existing container named $CONTAINER_NAME..."
enroot remove --force $CONTAINER_NAME || true

# Create new container
echo "Creating new container $CONTAINER_NAME..."
enroot create --name $CONTAINER_NAME /dss/dssfs02/lwp-dss-0001/pn76je/pn76je-dss-0000/tomasruiz/my_custom_pt.sqsh

# Start the container with necessary mounts and environment variables
enroot start --rw --root \
    --mount /dss/dsshome1/0D/di38bec/code:/workspace/code \
    --mount /dss/dsshome1/0D/di38bec/datasets:/workspace/datasets \
    --mount "$DSS_HOME:$DSS_HOME" \
    --env DSS_HOME=$DSS_HOME \
    --env HF_TOKEN=$HF_TOKEN \
    --env HF_HOME=$HF_HOME \
    --env VLLM_USE_V1=1 \
    $CONTAINER_NAME bash -c "
        source /root/ptvenv/bin/activate
        
        # Launch vLLM server in background
        echo 'Starting vLLM server...'
        vllm serve Qwen/Qwen3-32B \
          --speculative_config '{\"method\": \"draft_model\", \"model\": \"Qwen/Qwen3-1.7B\", \"num_speculative_tokens\": 3, \"max_model_len\": 20000, \"disable_padded_drafter_batch\": true}' \
          --max-model-len 20000 \
          --disable-uvicorn-access-log \
          --port $VLLM_PORT &
        
        SERVER_PID=\$!
        echo \"Server started with PID: \$SERVER_PID\"
        
        # Run benchmark (has built-in wait mechanism)
        echo 'Starting benchmark...'
        vllm bench serve \
          --model Qwen/Qwen3-32B \
          --dataset-name hf \
          --dataset-path philschmid/mt-bench \
          --num-prompts 80 \
          --max-concurrency 1 \
          --temperature 0.0 \
          --top-p 1.0 \
          --port $VLLM_PORT
        
        echo 'Benchmark completed, stopping server...'
        kill \$SERVER_PID
    "

# Clean up: remove the container after job completion
echo "Cleaning up: removing container $CONTAINER_NAME..."
enroot remove --force $CONTAINER_NAME || true
</details>

<img width="49%" alt="image" src="https://github.com/user-attachments/assets/bcfc012d-106a-449b-8fe3-94fc60795206" />
<img width="49%" alt="image" src="https://github.com/user-attachments/assets/d2f632ea-d9d3-4a6d-a2af-6520e9eb2882" />

tomasruizt commented:
Let me know what we could do to move this PR forward @benchislett @ekagra-ranjan @LiuXiaoxuanPKU @hmellor 🚀

ggg-s commented:
thank u. 
I noticed something strange: even after setting (and removing the TP restriction), the service starts fine, but the draft model still loads on every GPU. Looks like each TP rank is creating its own drafter instead of sharing a single TP=1 instance."draft_tensor_parallel_size": 1

ggg-s commented:
> > What is the TP you are using for Qwen3-32B? By default, draft model TP is equal to target model TP. Since Qwen3-1.7B is a small model, running it on high TP might be incurring nccl communication cost. Try setting draft TP to 1.
> 
> I ran the benchmarks with TP=1 and num_draft_tokens=3. So we can rule out TP communication issues.
11

QingNagi commented:
What's the achievable Mean Acceptance Length for this combination?


> 嗨，我目前正在使用 VLM 目标模型和 LLM 草稿模型测试推测解码。我注意到，当批量大小为 1 时，吞吐量提高了 14.7% 左右，但当批量大小大于 1 时，根本没有加速。我使用的数据集是纯文本（无图像）。我参考了您的 PR，并希望在此设置中看到加速，但它似乎没有生效。您能否分享一下为什么会发生这种情况的可能原因或见解？多谢！
> 
> vllm serve： CUDA_VISIBLE_DEVICES=0,1,2,3 vllm serve /models/Qwen2.5-VL-72B-AWQ --port 18000 --max-model-len 10000 --max-num-seqs 32 --tensor-parallel-size 4 --speculative-config '{ “method”： “draft_model”， “model”： “/models/Qwen2.5-1.5B”， “num_speculative_tokens”： 3， “max-model-len”： 10000， “disable_padded_drafter_batch”： true }'
> 
> 结果： <img alt="图像" width="1144" height="204" src="https://private-user-images.githubusercontent.com/71912051/500954515-a6b566cf-7f73-44b1-9070-e0cd374f078d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA0OTM1OTMsIm5iZiI6MTc2MDQ5MzI5MywicGF0aCI6Ii83MTkxMjA1MS81MDA5NTQ1MTUtYTZiNTY2Y2YtN2Y3My00NGIxLTkwNzAtZTBjZDM3NGYwNzhkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDE1VDAxNTQ1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM5MTBjYjQwZjZlZWIyYWE2MzBhMGMzNmEzNzFiMjYwYzFmZDRlNGI0OGFmZmNhYjMwNjFlOTZkNmI4M2EwYWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NrzDjnYseCscgFARcu45dAYOFJkUkH0eSyj95z3QX-0">

What's the achievable Mean Acceptance Length for this combination?

ggg-s commented:
it's 2.5

QingNagi commented:
> 是 2.5

This Mean Acceptance Length is okay, I think the draft_model method might not be suitable for high concurrency operations on this long input short output dataset.

ggg-s commented:
> I try it : My command: vllm serve Qwen3/Qwen3-32B-FP8 --served-model-name "qwen3" --gpu-memory-utilization 0.9 --tensor-parallel-size 2 --trust-remote-code --max-model-len 10000 --speculative-config '{"method": "draft_model", "model": "Qwen3/Qwen3-0.6B", "num_speculative_tokens": 5, "draft_tensor_parallel_size": 1, "max-model-len": 10000, "disable_padded_drafter_batch": true}' --override-generation-config '{"temperature" : 0}' & Error message: (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] WorkerProc failed to start. (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] Traceback (most recent call last): (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] File "/opt/vllm_env/lib/python3.10/site-packages/vllm/v1/executor/multiproc_executor.py", line 571, in worker_main (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] worker = WorkerProc(*args, **kwargs) (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] File "/opt/vllm_env/lib/python3.10/site-packages/vllm/v1/executor/multiproc_executor.py", line 437, in **init** (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] self.worker.load_model() (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] File "/opt/vllm_env/lib/python3.10/site-packages/vllm/v1/worker/gpu_worker.py", line 213, in load_model (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] self.model_runner.load_model(eep_scale_up=eep_scale_up) (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] File "/opt/vllm_env/lib/python3.10/site-packages/vllm/v1/worker/gpu_model_runner.py", line 2656, in load_model (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] self.drafter.load_model(self.model) (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] File "/opt/vllm_env/lib/python3.10/site-packages/vllm/v1/spec_decode/draft_model.py", line 147, in load_model (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] self.model = get_model( (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] File "/opt/vllm_env/lib/python3.10/site-packages/vllm/model_executor/model_loader/**init**.py", line 122, in get_model (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] return loader.load_model( (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] File "/opt/vllm_env/lib/python3.10/site-packages/vllm/model_executor/model_loader/base_loader.py", line 50, in load_model (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] self.load_weights(model, model_config) (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] File "/opt/vllm_env/lib/python3.10/site-packages/vllm/model_executor/model_loader/default_loader.py", line 276, in load_weights (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] raise ValueError("Following weights were not initialized from " (Worker_TP0 pid=112985) ERROR 10-13 12:01:14 [multiproc_executor.py:597] ValueError: Following weights were not initialized from checkpoint: {'model.layers.18.self_attn.attn.prob_scale', 'model.layers.7.self_attn.attn.k_scale', 'model.layers.26.self_attn.attn.v_scale', 'model.layers.0.mlp.gate_up_proj.weight_scale_inv', 'model.layers.27.self_attn.attn.v_scale', 'model.layers.0.self_attn.attn.q_scale', 'model.layers.27.self_attn.qkv_proj.weight_scale_inv', 'model.layers.9.self_attn.attn.q_scale', 'model.layers.27.mlp.down_proj.weight_scale_inv', 'model.layers.2.mlp.gate_up_proj.weight_scale_inv', 'model.layers.12.mlp.down_proj.weight_scale_inv', 'model.layers.23.self_attn.o_proj.weight_scale_inv', 'model.layers.25.self_attn.qkv_proj.weight_scale_inv', 'model.layers.5.self_attn.attn.v_scale', 'model.layers.14.self_attn.attn.k_scale', 'model.layers.18.self_attn.qkv_proj.weight_scale_inv', 'model.layers.20.self_attn.attn.q_scale', 'model.layers.12.self_attn.attn.k_scale', 'model.layers.18.self_attn.attn.k_scale', 'model.layers.9.self_attn.o_proj.weight_scale_inv', 'model.layers.8.self_attn.qkv_proj.weight_scale_inv', 'model.layers.22.self_attn.attn.k_scale', 'model.layers.25.self_attn.attn.k_scale', 'model.layers.20.self_attn.o_proj.weight_scale_inv', 'model.layers.19.mlp.down_proj.weight_scale_inv', 'model.layers.9.self_attn.attn.prob_scale', 'model.layers.24.mlp.gate_up_proj.weight_scale_inv', 'model.layers.26.mlp.gate_up_proj.weight_scale_inv', 'model.layers.17.self_attn.o_proj.weight_scale_inv', 'model.layers.0.mlp.down_proj.weight_scale_inv', 'model.layers.4.self_attn.o_proj.weight_scale_inv', 'model.layers.4.self_attn.attn.v_scale', 'model.layers.6.self_attn.o_proj.weight_scale_inv', 'model.layers.7.self_attn.attn.v_scale', 'model.layers.16.mlp.down_proj.weight_scale_inv', 'model.layers.7.mlp.gate_up_proj.weight_scale_inv', 'model.layers.1.self_attn.o_proj.weight_scale_inv', 'model.layers.0.self_attn.attn.prob_scale', 'model.layers.27.self_attn.attn.prob_scale', 'model.layers.21.self_attn.attn.q_scale', 'model.layers.19.self_attn.attn.k_scale', 'model.layers.27.self_attn.o_proj.weight_scale_inv', 'model.layers.4.self_attn.attn.prob_scale', 'model.layers.2.self_attn.qkv_proj.weight_scale_inv', 'model.layers.10.mlp.down_proj.weight_scale_inv', 'model.layers.9.self_attn.attn.k_scale', 'model.layers.24.self_attn.attn.v_scale', 'model.layers.15.self_attn.attn.q_scale', 'model.layers.21.self_attn.o_proj.weight_scale_inv', 'model.layers.12.self_attn.attn.v_scale', 'model.layers.21.mlp.gate_up_proj.weight_scale_inv', 'model.layers.15.self_attn.attn.prob_scale', 'model.layers.12.self_attn.qkv_proj.weight_scale_inv', 'model.layers.11.self_attn.attn.q_scale', 'model.layers.24.self_attn.o_proj.weight_scale_inv', 'model.layers.12.self_attn.attn.q_scale', 'model.layers.10.self_attn.attn.q_scale', 'model.layers.3.self_attn.attn.q_scale', 'model.layers.7.mlp.down_proj.weight_scale_inv', 'model.layers.5.self_attn.o_proj.weight_scale_inv', 'model.layers.27.self_attn.attn.k_scale', 'model.layers.21.self_attn.attn.v_scale', 'model.layers.20.self_attn.attn.v_scale', 'model.layers.17.self_attn.attn.k_scale', 'model.layers.14.self_attn.qkv_proj.weight_scale_inv', 'model.layers.3.self_attn.o_proj.weight_scale_inv', 'model.layers.5.self_attn.attn.prob_scale', 'model.layers.13.mlp.down_proj.weight_scale_inv', 'model.layers.11.mlp.gate_up_proj.weight_scale_inv', 'model.layers.13.self_attn.attn.v_scale', 'model.layers.1.self_attn.attn.v_scale', 'model.layers.13.self_attn.qkv_proj.weight_scale_inv', 'model.layers.0.self_attn.attn.v_scale', 'model.layers.11.self_attn.attn.prob_scale', 'model.layers.4.mlp.gate_up_proj.weight_scale_inv', 'model.layers.7.self_attn.attn.q_scale', 'model.layers.17.self_attn.attn.q_scale', 'model.layers.20.self_attn.qkv_proj.weight_scale_inv', 'model.layers.4.mlp.down_proj.weight_scale_inv', 'model.layers.6.self_attn.attn.v_scale', 'model.layers.11.self_attn.o_proj.weight_scale_inv', 'model.layers.18.self_attn.attn.v_scale', 'model.layers.20.mlp.gate_up_proj.weight_scale_inv', 'model.layers.10.self_attn.qkv_proj.weight_scale_inv', 'model.layers.19.self_attn.attn.v_scale', 'model.layers.24.self_attn.attn.k_scale', 'model.layers.26.self_attn.attn.prob_scale', 'model.layers.0.self_attn.qkv_proj.weight_scale_inv', 'model.layers.9.self_attn.qkv_proj.weight_scale_inv', 'model.layers.23.self_attn.attn.v_scale', 'model.layers.18.mlp.down_proj.weight_scale_inv', 'model.layers.14.self_attn.attn.prob_scale', 'model.layers.5.mlp.down_proj.weight_scale_inv', 'model.layers.26.self_attn.attn.k_scale', 'model.layers.23.mlp.down_proj.weight_scale_inv', 'model.layers.15.self_attn.attn.k_scale', 'model.layers.3.mlp.gate_up_proj.weight_scale_inv', 'model.layers.9.mlp.down_proj.weight_scale_inv', 'model.layers.22.self_attn.qkv_proj.weight_scale_inv', 'model.layers.12.mlp.gate_up_proj.weight_scale_inv', 'model.layers.22.self_attn.o_proj.weight_scale_inv', 'model.layers.24.mlp.down_proj.weight_scale_inv', 'model.layers.24.self_attn.qkv_proj.weight_scale_inv', 'model.layers.20.mlp.down_proj.weight_scale_inv', 'model.layers.20.self_attn.attn.prob_scale', 'model.layers.1.self_attn.attn.prob_scale', 'model.layers.22.self_attn.attn.prob_scale', 'model.layers.19.self_attn.qkv_proj.weight_scale_inv', 'model.layers.17.self_attn.qkv_proj.weight_scale_inv', 'model.layers.16.self_attn.attn.v_scale', 'model.layers.3.self_attn.attn.v_scale', 'model.layers.5.self_attn.attn.q_scale', 'model.layers.14.mlp.down_proj.weight_scale_inv', 'model.layers.10.self_attn.attn.prob_scale', 'model.layers.27.mlp.gate_up_proj.weight_scale_inv', 'model.layers.1.self_attn.attn.q_scale', 'model.layers.6.mlp.gate_up_proj.weight_scale_inv', 'model.layers.20.self_attn.attn.k_scale', 'model.layers.22.mlp.down_proj.weight_scale_inv', 'model.layers.2.self_attn.attn.v_scale', 'model.layers.8.self_attn.o_proj.weight_scale_inv', 'model.layers.13.self_attn.o_proj.weight_scale_inv', 'model.layers.2.self_attn.attn.prob_scale', 'model.layers.14.self_attn.o_proj.weight_scale_inv', 'model.layers.21.mlp.down_proj.weight_scale_inv', 'model.layers.17.self_attn.attn.v_scale', 'model.layers.8.self_attn.attn.k_scale', 'model.layers.0.self_attn.attn.k_scale', 'model.layers.4.self_attn.qkv_proj.weight_scale_inv', 'model.layers.2.self_attn.o_proj.weight_scale_inv', 'model.layers.22.mlp.gate_up_proj.weight_scale_inv', 'model.layers.21.self_attn.attn.prob_scale', 'model.layers.9.self_attn.attn.v_scale', 'model.layers.4.self_attn.attn.q_scale', 'model.layers.19.self_attn.attn.prob_scale', 'model.layers.8.mlp.gate_up_proj.weight_scale_inv', 'model.layers.5.mlp.gate_up_proj.weight_scale_inv', 'model.layers.12.self_attn.o_proj.weight_scale_inv', 'model.layers.7.self_attn.qkv_proj.weight_scale_inv', 'model.layers.8.self_attn.attn.v_scale', 'model.layers.13.self_attn.attn.prob_scale', 'model.layers.13.self_attn.attn.q_scale', 'model.layers.17.mlp.down_proj.weight_scale_inv', 'model.layers.16.self_attn.attn.q_scale', 'model.layers.19.mlp.gate_up_proj.weight_scale_inv', 'model.layers.24.self_attn.attn.prob_scale', 'model.layers.12.self_attn.attn.prob_scale', 'model.layers.19.self_attn.attn.q_scale', 'model.layers.26.self_attn.o_proj.weight_scale_inv', 'model.layers.2.mlp.down_proj.weight_scale_inv', 'model.layers.25.self_attn.attn.q_scale', 'model.layers.25.self_attn.o_proj.weight_scale_inv', 'model.layers.9.mlp.gate_up_proj.weight_scale_inv', 'model.layers.3.self_attn.attn.k_scale', 'model.layers.4.self_attn.attn.k_scale', 'model.layers.23.mlp.gate_up_proj.weight_scale_inv', 'model.layers.11.self_attn.attn.k_scale', 'model.layers.0.self_attn.o_proj.weight_scale_inv', 'model.layers.13.mlp.gate_up_proj.weight_scale_inv', 'model.layers.18.self_attn.o_proj.weight_scale_inv', 'model.layers.2.self_attn.attn.k_scale', 'model.layers.5.self_attn.qkv_proj.weight_scale_inv', 'model.layers.1.mlp.down_proj.weight_scale_inv', 'model.layers.25.mlp.gate_up_proj.weight_scale_inv', 'model.layers.26.self_attn.attn.q_scale', 'model.layers.26.self_attn.qkv_proj.weight_scale_inv', 'model.layers.15.mlp.down_proj.weight_scale_inv', 'model.layers.22.self_attn.attn.v_scale', 'model.layers.27.self_attn.attn.q_scale', 'model.layers.22.self_attn.attn.q_scale', 'model.layers.23.self_attn.attn.k_scale', 'model.layers.10.mlp.gate_up_proj.weight_scale_inv', 'model.layers.2.self_attn.attn.q_scale', 'model.layers.16.self_attn.attn.k_scale', 'model.layers.10.self_attn.attn.v_scale', 'model.layers.16.mlp.gate_up_proj.weight_scale_inv', 'model.layers.14.self_attn.attn.q_scale', 'model.layers.23.self_attn.attn.prob_scale', 'model.layers.1.mlp.gate_up_proj.weight_scale_inv', 'model.layers.25.self_attn.attn.prob_scale', 'model.layers.7.self_attn.attn.prob_scale', 'model.layers.23.self_attn.attn.q_scale', 'model.layers.15.mlp.gate_up_proj.weight_scale_inv', 'model.layers.10.self_attn.o_proj.weight_scale_inv', 'model.layers.19.self_attn.o_proj.weight_scale_inv', 'model.layers.18.mlp.gate_up_proj.weight_scale_inv', 'model.layers.16.self_attn.qkv_proj.weight_scale_inv', 'model.layers.23.self_attn.qkv_proj.weight_scale_inv', 'model.layers.14.mlp.gate_up_proj.weight_scale_inv', 'model.layers.25.self_attn.attn.v_scale', 'model.layers.11.mlp.down_proj.weight_scale_inv', 'model.layers.10.self_attn.attn.k_scale', 'model.layers.14.self_attn.attn.v_scale', 'model.layers.3.self_attn.qkv_proj.weight_scale_inv', 'model.layers.11.self_attn.qkv_proj.weight_scale_inv', 'model.layers.11.self_attn.attn.v_scale', 'model.layers.13.self_attn.attn.k_scale', 'model.layers.3.self_attn.attn.prob_scale', 'model.layers.17.mlp.gate_up_proj.weight_scale_inv', 'model.layers.7.self_attn.o_proj.weight_scale_inv', 'model.layers.26.mlp.down_proj.weight_scale_inv', 'model.layers.6.self_attn.attn.k_scale', 'model.layers.8.self_attn.attn.q_scale', 'model.layers.8.self_attn.attn.prob_scale', 'model.layers.3.mlp.down_proj.weight_scale_inv', 'model.layers.6.self_attn.qkv_proj.weight_scale_inv', 'model.layers.6.self_attn.attn.q_scale', 'model.layers.16.self_attn.o_proj.weight_scale_inv', 'model.layers.18.self_attn.attn.q_scale', 'model.layers.8.mlp.down_proj.weight_scale_inv', 'model.layers.15.self_attn.qkv_proj.weight_scale_inv', 'model.layers.15.self_attn.o_proj.weight_scale_inv', 'model.layers.5.self_attn.attn.k_scale', 'model.layers.6.self_attn.attn.prob_scale', 'model.layers.21.self_attn.attn.k_scale', 'model.layers.25.mlp.down_proj.weight_scale_inv', 'model.layers.16.self_attn.attn.prob_scale', 'model.layers.6.mlp.down_proj.weight_scale_inv', 'model.layers.21.self_attn.qkv_proj.weight_scale_inv', 'model.layers.24.self_attn.attn.q_scale', 'model.layers.1.self_attn.qkv_proj.weight_scale_inv', 'model.layers.1.self_attn.attn.k_scale', 'model.layers.17.self_attn.attn.prob_scale', 'model.layers.15.self_attn.attn.v_scale'}

Hi, I noticed that even though I set "draft_tensor_parallel_size": 1, both the logs and GPU memory usage show that the draft model is still being loaded on every GPU.

Could you confirm whether this command actually supports running the target model with TP=2 and the draft model with TP=1?

QingNagi commented:
> I think I know precisely the line causing it. But to reproduce I would need the questions.jsonl file. Would it be possible to share it? @QingNagi

Sure! Download the dataset using: wget https://raw.githubusercontent.com/hemingkx/Spec-Bench/refs/heads/main/data/spec_bench/question.jsonl

tomasruizt commented:
@QingNagi can you share what commit you are testing on?

tomasruizt commented:
@QingNagi I meant which commit of this branch?

tomasruizt commented:
@QingNagi Are you on the vLLM Slack? We can chat over there

ggg-s commented:
@tomasruizt I was wondering if this PR ([#26937](https://github.com/vllm-project/vllm/pull/26937)) could help improve the decoding speed of the draft model?

ggg-s commented:
@tomasruizt Can I contact you on Slack?

tomasruizt commented:
@LiuXiaoxuanPKU I ran throughput benchmarks on a multi-GPU (TP=4) setup with `Llama-3-70B`, and added the results to the bottom of the PR description. I found speedups up to batch_size=32. AFAIK you had a somewhat different results. If you can share those details, I can attempt to reproduce. Thanks for your review :)

tomasruizt commented:
@benchislett I added the EAGLE3 benchmark for reference. It shows very good acceleration, strongest with small batch sizes. It's faster than the draft_model contributed here, but using draft_model doesn't require training a separate model, like EAGLE. Furthermore, there are follow-up optimizations that should substantially improve draft_model performance:
* (1) Making the draft model run on TP=1, regardless of the TP of the target model. The Qwen3 benchmarks on TP=1 in the PR description show that we might reach 2x acceleration this way.
* (2) the draft model should run with full CUDA graphs during pure decode steps (just like the target model). Profiling suggest that this could speed up draft_model decodes by ~20-30%.

I tried both for this PR, but found them non-trivial, so I stepped back. I will be away in November, so I think we should try to merge this PR before November.

benchislett commented:
Here are the blockers in my opinion:
- Doesn't integrate with padded-batch spec decoding, simply due to implementation choices. Since we're hoping to use padded-batch more and more going forward, it's important that support be included.
- "extend" logic is hard to read and isn't very efficient. Causes a non-trivial bottleneck in input preparation according to my benchmarking. I have a custom kernel that fuses this logic but I think it is more generally applicable and I'm not sure how I want to scope it in the context of this PR. I think we can find a better way to implement the extending logic.
- Doesn't integrate perfectly in the current specdec framework. While this branch does reuse code with EAGLE, it still muddles a lot of the responsibilities. I think it needs a thorough refactoring pass to properly separate concerns here and find the right trade-off between complexity and code reuse. This may involve a preliminary PR to refactor vLLM's current spec decoding interfaces before we can contribute this code.

ofirzaf commented:
> This may involve a preliminary PR to refactor vLLM's current spec decoding interfaces before we can contribute this code.

@benchislett What are the issues in the current interface in your opinion?
To me it seemed like Eagle was implemented first and then all the other methods were adjusted to follow the same flow even if it doesn't make much sense. I think there is a problem where different methods have a different flows.
Anyway, I would like to help shaping this

tomasruizt commented on vllm/v1/attention/backends/utils.py:
```diff
@@ -187,6 +194,43 @@ def slice_query_start_locs(
+        block_table_tensor=cad.block_table_tensor,
+        slot_mapping=new_slot_mapping,
+    )
+    return new_cad
```
This comment is duplicated

benchislett commented on vllm/config/speculative.py:
```diff
@@ -74,6 +74,9 @@ class SpeculativeConfig:
     or the same as the target model's tensor parallel size."""
+    tensor_parallel_size: int | None = None
+    """Users should pass "draft_tensor_parallel_size". This parameters is only 
+    to reject it if passed."""
```
```suggestion
    """Users should pass "draft_tensor_parallel_size". This parameter's purpose is to warn
    users when they mistakenly provide the wrong argument."""
```

benchislett commented on vllm/v1/attention/backends/utils.py:
```diff
@@ -96,6 +96,16 @@ class CommonAttentionMetadata:
+        return self.seq_lens.shape[0]
+
+    def naive_query_lens(self) -> torch.Tensor:
+        """Naive because it assumes that query ends where the next query starts."""
```
Why would this ever _not_ be the case? This is how we compute `query_lens` pretty much everywhere.

benchislett commented on vllm/v1/spec_decode/metrics.py:
```diff
@@ -214,6 +215,24 @@ def observe(self, spec_decoding_stats: SpecDecodingStats, engine_idx: int = 0):
             counter.inc(spec_decoding_stats.num_accepted_tokens_per_pos[pos])
 
 
+def compute_acceptance_rate(metrics: list[Metric]) -> float:
```
Where are these used?

benchislett reviewed (approved):
Excellent work! The code looks great, I'm very pleased with the current state. Left a few more comments, mostly looking for clarification and not requesting changes.

Thanks!

benchislett commented:
Another question: do you know if this will work for draft models with a different attention setup as the target model? For instance, Qwen3-4B has 32 attention heads per layer and Qwen3 235B A22B has 64 attention heads per layer. Is this expected to work somehow? What happens if you try to launch vLLM with this setup, does it crash with a meaningful error message?

tomasruizt commented:
> Another question: do you know if this will work for draft models with a different attention setup as the target model? For instance, Qwen3-4B has 32 attention heads per layer and Qwen3 235B A22B has 64 attention heads per layer. Is this expected to work somehow? What happens if you try to launch vLLM with this setup, does it crash with a meaningful error message?

I haven't checked this case. Will do, and post the results in this comment.

yewentao256 commented on vllm/v1/spec_decode/draft_model.py:
```diff
@@ -0,0 +1,271 @@
+            get_layers_from_vllm_config(self.vllm_config, Attention).keys()
+            - target_attn_layer_names
+        )
+        self.attn_layer_names = list(draft_attn_layer_names)
```
Note that list of set may lose order, potential bug if the order counts in later usage

yewentao256 reviewed (commented):
Thanks for the work! A few thoughts

tomasruizt commented on vllm/v1/spec_decode/draft_model.py:
```diff
@@ -0,0 +1,271 @@
+        target_token_ids: torch.Tensor,
+        next_token_ids: torch.Tensor,
+        target_positions: torch.Tensor,
+        last_token_indices: torch.Tensor | None,
```
The param is unused by `DraftModelProposer`, but needed by `SpecDecodeBaseProposer` for EAGLE

tomasruizt committed:
Add speculative decoding with draft model support

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

tomasruizt committed:
Relax AL and AR thresholds in quantization tests

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

tomasruizt committed:
Fix prefix error in test_bind_kv_cache_draft_model()

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

tomasruizt committed:
Raise NotImplementedError when speculative method=None

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

tomasruizt committed:
Improve readability in if-else

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

tomasruizt committed:
Use dtype int32 in torch.arange()

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

tomasruizt committed:
Extend CUDA graph compile ranges by num speculative tokens

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

tomasruizt committed:
Added test triggering failure

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

tomasruizt committed:
Update variable name

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

tomasruizt committed:
Reduce the range extension to only max_num_seqs

Signed-off-by: Tomas Ruiz <tomas.ruiz.te@gmail.com>

micah-wil commented on tests/v1/e2e/test_spec_decode.py:
```diff
@@ -583,3 +614,269 @@ def test_mtp_correctness(
+    n_accepted_toks = name2metric["vllm:spec_decode_num_accepted_tokens"].value  # type: ignore
+    if n_drafts == 0:
+        return 1
+    return 1 + (n_accepted_toks / n_drafts)
```
Gotcha, thanks, I am consistently getting 3.89 on MI325. Maybe I'll lower `"expected_acceptance_len"` to 3.6 just to be safe?

tomasruizt replied:
That's a big drop, I suggest you drop it only to 3.85 or 3.8

micah-wil replied:
Ok, let's go with 3.8 then if that's alright https://github.com/vllm-project/vllm/pull/32731
